File: NumberTheory/MillerRabin.py
import random

def witness(a, n):
    t, u = 0, n - 1
    while u % 2 == 0:
        t += 1
        u >>= 1
    x, _x = pow(a, u, n), 0
    while t:
        _x = (x * x) % n
        if _x == 1 and x != 1 and x != n - 1:
            return True
        x = _x
        t -= 1
    return _x != 1

def isPrime(x):
    if x < 2:
        return False
    for i in {2, 3, 5, 7, 11, 13, 17, 19}:
        if x % i == 0:
            return x == i
    for i in range(10):
        if witness(random.randint(1, x), x):
            return False
    return True


for i in range(1, 1000):
    if isPrime((2 ** i) + 1):
        print(i, (2 ** i) + 1)

File: NumberTheory/Miller_Rabin.cpp
typedef long long ll;

const int psize = 1010000;
bool isp[psize];
int prime[psize], tot;

void prime_table() {
  register int i, j;
  for (i = 2, tot = 0; i < psize; i++) {
    if (!isp[i]) prime[tot++] = i;
    for (j = 0; j < tot && prime[j] * i < psize; j++) {
      isp[prime[j] * i] = true;
      if (i % prime[j] == 0) break;
    }
  }
}

ll qpow(ll a, ll t, ll mod) {
  ll b = 1;
  for (; t > 0; t >>= 1, a = a * a % mod) {
    if (t & 1) {
      b = b * a % mod;
    }
  }
  return b;
}

bool witness(ll a, ll n) {
  int t = 0;
  ll u = n - 1;
  for (; ~u & 1; u >>= 1) t++;
  ll x = qpow(a, u, n), _x = 0;
  while (t--) {
    _x = mul(x, x, n);
    if (_x == 1 && x != 1 && x != n - 1) return true;
    x = _x;
  }
  return _x != 1;
}

// WA -> make `t` higher
// TLE -> make `t` lower
bool isprime(ll n, int t=6) {
  if (n < 2) return false;
  if (n < psize) return !isp[n];
  if (~n & 1) return false;
  for (int j = 0; j < t; j++) {
    if (witness(rand() % (n - 1) + 1, n)) {
      return false;
    }
  }
  return true;
}
